Geometric sum evaluates correctly for float base(Fix for : #11642)

ArifAhmed1995 opened the pull request:
Reference : #11642 

Now geometric sum evaluates correctly for both Rational as well as float bases
Example : 

```
>>Sum((0.35)**(n), (n, 1, oo)).doit()
0.538461538461538
>>Sum((1.35)**(n), (n, 1, oo)).doit()
oo
>>Sum((0.5)**(n), (n, 1, oo)).doit()
1.00000000000000
```

Some unit tests have also been added.

ArifAhmed1995 committed:
matrices:Matrix**exponent stays unevaluated for non-integer exponent

If exponent can be sympified to a sympy integer then matrix**exponent
is evaluated by default.Otherwise , it is stored in a MatPow object
and can be evaluated by calling the doit() method.

Some changes have also been done to other test functions which
earlier evaluated Rational and symbolic exponents by default.

Another new feature has been added : doit() on a MatPow object
evaluates for symbolic exponents as well.

ArifAhmed1995 committed:
Removed uneccessary whitespaces

ArifAhmed1995 committed:
MatPow raised to -1 should have -1 clubbed with previous exponent

ArifAhmed1995 committed:
Merge branch 'master' of https://github.com/sympy/sympy into matrices

ArifAhmed1995 committed:
Added requested changes

ArifAhmed1995 committed:
Add requested changes

ArifAhmed1995 committed:
Removed uneccessary lines from git merge

ArifAhmed1995 committed:
summations : Geometric sum correctly evaluates for float base

doit() for geometric sums now evaluates for float base too.
Example :Previously , Sum((0.5)**(n), (n, 1, oo)).doit()
	 would not evaluate ;now it evaluates to 1.00000000

ArifAhmed1995 committed:
Removed extra whitespaces

smichr commented on sympy/concrete/tests/test_sums_products.py:
```diff
@@ -255,6 +255,12 @@ def test_geometric_sums():
     assert Sum(1/(n**3 - 1), (n, -oo, -2)).doit() == summation(1/(n**3 - 1), (n, -oo, -2))
 
+    #issue 11642:
+    assert Sum(0.5**n,(n, 1, oo)).doit() == 1.0
```
I suspect this comes back as `1`, not `1.0`. I think it should come back as a Float: if the input is not exact, the output should not be exact. You can test the result by `assert result.is_Float` in addition to testing the value (`result == 1`).

smichr commented on sympy/matrices/expressions/matexpr.py:
```diff
@@ -127,12 +127,20 @@ def __pow__(self, other):
         elif self.is_Identity:
             return self
-        elif other is S.NegativeOne:
+        elif other is S.NegativeOne and not isinstance(self,MatPow):
```
space after commas here and below

smichr commented on sympy/matrices/expressions/tests/test_matpow.py:
```diff
@@ -43,7 +43,7 @@ def test_as_explicit():
     # less expensive than testing on a 2x2
     A = ImmutableMatrix([4]);
-    assert MatPow(A, S.Half).as_explicit() == A**S.Half
+    assert MatPow(A, S.Half).as_explicit() == (A**S.Half ).doit()
```
no space before `)`

smichr commented on sympy/matrices/matrices.py:
```diff
@@ -208,16 +207,15 @@ def __pow__(self, num):
-                                if isinstance(bn, binomial):
-                                        bn = bn._eval_expand_func()
-                                jc[j, i+j] = l**(n-i)*bn
+                    for j in range(N-i):
```
space around `-`

smichr commented on sympy/matrices/matrices.py:
```diff
@@ -226,8 +224,20 @@ def jordan_cell_power(jc, n):
         else:
-            raise TypeError(
-                "Only SymPy expressions or int objects are supported as exponent for matrices")
+            raise TypeError("Only SymPy expressions or int objects are supported as exponent for matrices")
```
"supported as matrix exponents."

smichr commented on sympy/matrices/tests/test_matrices.py:
```diff
@@ -196,24 +196,29 @@ def test_power():
 
     A = Matrix([[33, 24], [48, 57]])
-    assert (A**(S(1)/2))[:] == [5, 2, 4, 7]
+    assert ( (A**(S(1)/2)).doit() )[:] == [5, 2, 4, 7]
```
remove space after/before `)`

ArifAhmed1995 committed:
Add requested changes

ArifAhmed1995 commented:
@smichr The commit history is terribly messed up as two series of commits from separate fixes got intermingled.I've deleted the branch and closed the PR.

I have implemented the changes requested in two separate PRs : #11682 and #11648 . Can you look through the changes now?
